server/src/main/java/org/elasticsearch/action/search/SearchResponse.java

```diff
@@ -570,29 +627,44 @@ public int getTotal() {
          * @return how many total clusters the search was executed successfully on
          */
         public int getSuccessful() {
-            return successful;
+            if (clusterInfo.isEmpty()) {
+                return successful;
+            } else {
+                return (int) clusterInfo.values()
+                    .stream()
+                    .filter(c -> c.getStatus() == Cluster.Status.SUCCESSFUL || c.getStatus() == Cluster.Status.PARTIAL)
+                    .count();
+            }
```

Comment:
As mentioned above, from a thread-safety and accuracy point of view, this is best effort, not guaranteed accurate as the Cluster objects might in the act of being updated while a "status" update query is reading the Cluster object state. For performance reasons, this is considered a fair tradeoff, since the final status is the only one guaranteed to be truly accurate.